---
title: "Ribomethseq QC report"
output:
  rmdformats::downcute
---

```{r, echo=FALSE}
library(ggpubr)
knitr::opts_chunk$set(echo=FALSE)
knitr::opts_chunk$set(message=FALSE)
```


Date : `r Sys.Date()`

# Overview

```{r,fig.height = 10, fig.width = 10, fig.align = "center"}
  table_sorted = total_table[c("Filename","%Surviving","%Used_reads_for_counting")] 
  
  colnames(table_sorted) = c("Sample","% of reads kept by Trimmomatic","% of reads aligned by bowtie2")
  d <- melt(table_sorted, id.vars="Sample")
  
  sample_order = reorder(d$Sample,d$value)
plotTrimBowtie(d,sample_order)
plotSummaryNewFlags(total_table,fdl,sample_order)
```

# First analysis (before trimmomatic)

## Unique vs duplicated reads for each sample

The following plots show the raw number and percentage of reads that are either unique or duplicated.
In ribomethseq, it's not uncommon to have a lot of duplicated reads. This can be wrongly flagged by FastQC as being problematic.

### Raw
```{r,fig.height = 10, fig.width = 10, fig.align = "center"}
plotReadTotals(fdl, xlim=sample_order)
```

### Percentage
```{r}
plotReadTotalsPercentage(ggplot_total_table, "Filename", "PUnique_Reads", "PDuplicated_Reads", sample.order = sample_order)
```

## Base Sequence Content {#bsc}

Per Base Sequence Content plots out the proportion of each base position in a file for which each of the four normal DNA bases has been called.
[more explanations](#bsc_indepth)

```{r}
plotSeqContent(fdl)
```

## N content par base 
```{r}
plotNContent(fdl)
```


## Mean quality of sequence by sample 

The mean quality of sequence/read by sample is evaluated through the phred score. The higher the better.

Reads with a low phred score will be excluded by Trimmomatic.

```{r,fig.height = 10, fig.width = 10, fig.align = "center"}
plotSeqQuals(fdl, plotType = "line")
```

## GC Content
```{r gc_content,fig.height = 10, fig.width = 10, fig.align = "center"}
plotGcContent(fdl, plotType = "line", theoreticalGC = F)
```


## Overrepresented sequences (FastQC)

```{r overrepresented_sequences}
plotOverrep(fdl)
```



# Trimmomatic

Trimmomatic is used in this pipeline to clean our sample from adapter content and bad reads. Reads can also be trimmed.

[More explanations about trimmomatic here](#trim_explanation)
[Check trimmomatic parameters here](#trim_parameters)

## Dropped reads (Trimmomatic)

Trimmomatic will exclude reads if :

* They are shorter than the minimum length specified by the user (check MINLEN parameter)
* They have an average quality lower than a value specified by the user (check AVGQUAL parameter)
* 

```{r trimmomatic_plot}
plot1 = plotTrimmomaticDropped(total_table, sample.order = sample_order)
plot2 = plotTrimmomaticDropped(total_table, percentage = T, sample.order = sample_order)
ggarrange(plot1, plot2, ncol = 2, common.legend = T)
```


# Sequences alignment (Bowtie 2)

## Aligned reads

The following plots (raw and percentage) show how many reads were successfully (or not) aligned one or multiple times. 
Unaligned reads are discarded. 

```{r aligned_reads}
plot1 = plotBowtieReport(bowtie_logs, sample.order = sample_order)
plot2 = plotBowtieReport(bowtie_logs, percentage = T, sample.order = sample_order)
ggarrange(plot1, plot2, ncol = 2, common.legend = T)
```

# Annexes

## The csv table {#csv_table}
The csv file is exported alongside this html.
You can check each column meaning [here](#table_indepth)

```{r}
library(DT)
datatable(total_table)
```

## How the pipeline works


```{r add_image, echo=FALSE, fig.cap="How the pipeline works", out.width = '100%'}
pic_path = paste(input_installDir,"/pipeline.png",sep="")
knitr::include_graphics(pic_path)
```

### FastQC

The purpose of FastQC is to give a detailed report for each sample before they go through the pipeline. The fastQ are not modified during this step.

### Trimmomatic {#trim_explanation}

Trimmomatic has three main purposes :

* Checks average quality and drops the whole read if it's below a specified value (AVGQUALITY parameter)
* Remove Illumina adapter (ILLUMINACLIP parameter)
* Trim both ends if the base quality is below a specified value (LEADING and TRAILING parameters)

In theory, all samples should be cleaned off bad reads and adapter content. 

### Bowtie 2

Bowtie 2 will align the reads of each sample against a reference. It will give an alignment file that will be used for the "counting" step. Unaligned reads are discarded.

## More in depth

### The table {#table_indepth}

[return to the result](#csv_table)

Below, you will find the explanation for each column, and from which tool it is from:

From FastQC :

* Filename : (shortened) name of the file
* Total_sequences : the number of reads inside the file
* Sequences_flagged_as_poor_quality : FastQC's verdict on the sequences inside the file. 
* shortest_sequence : length in bp of the shortest_sequence in the file
* longest_sequence : length in bp of the longest_sequence in the file
* %GC : percentage of GC inside the reads
* File type : 
* Encoding : how the files were encoded
* Total_deduplicated_percentage : the percentage of deduplicated sequences

From Trimmomatic :

* duplicated_reads : the number of duplicated reads inside this file (check PDuplicated_reads for the percentage of duplicated reads among total reads)
* Unique_read : the number of unique read (check PUnique_reads for the percentage of unique reads among total reads)
* Dropped : the number of dropped reads (check PDropped for the percentage of dropped reads among total reads)

From bowtie 2 :

* unaligned : number/percentage of reads that were not aligned by bowtie 2. They are discarded for the rest of the analysis.
* unique_unpaired : number/percentage of unpaired reads that have been aligned only one time
* multiple_unpaired : number/percentage of unpaired reads that have been aligned multiple times

### Per base sequence content {#bsc_indepth}

[return to results](#bsc)

In a random library you would expect that there would be little to no difference between the different bases of a sequence run, so the lines in this plot should run parallel with each other. The relative amount of each base should reflect the overall amount of these bases in your genome, but in any case they should not be hugely imbalanced from each other.

It's worth noting that some types of library will always produce biased sequence composition, normally at the start of the read. Libraries produced by priming using random hexamers (including nearly all RNA-Seq libraries) and those which were fragmented using transposases inherit an intrinsic bias in the positions at which reads start. This bias does not concern an absolute sequence, but instead provides enrichement of a number of different K-mers at the 5' end of the reads. Whilst this is a true technical bias, it isn't something which can be corrected by trimming and in most cases doesn't seem to adversely affect the downstream analysis. It will however produce a warning or error in this module. 

### overrepresented sequences

How FastQC works:

A normal high-throughput library will contain a diverse set of sequences, with no individual sequence making up a tiny fraction of the whole. Finding that a single sequence is very overrepresented in the set either means that it is highly biologically significant, or indicates that the library is contaminated, or not as diverse as you expected.

For each overrepresented sequence the program will look for matches in a database of common contaminants and will report the best hit it finds. Hits must be at least 20bp in length and have no more than 1 mismatch. Finding a hit doesn't necessarily mean that this is the source of the contamination, but may point you in the right direction. It's also worth pointing out that many adapter sequences are very similar to each other so you may get a hit reported which isn't technically correct, but which has very similar sequence to the actual match.

source : https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html

**In ribomethseq, it is not uncommon to have overrepresented sequences. This will most likely issue a "fail" in FastQC test.**


## Technical details

### R Session
```{r}
sessionInfo()
```

### Trimmomatic

version:
```{bash getTrimVersion}
trimmomatic -version
```

used parameters:
```{r}
datatable(trim_parameters)
```

"Not specified" means that the user has not specifed a value for this parameter.

### Bowtie 2
version:
```{bash getBowtieVersion}
bowtie2 --version
```

used parameters:
